contrib/babelfishpg_tsql/src/tsql_for/forxml.c

```diff
@@ -0,0 +1,329 @@
+			 * value being processed, but relation id/name is not provided by aggregate functions. We need to make
+			 * relation id available in aggregate functions in order to support AUTO mode.
+			 */
+			ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("AUTO mode is not supported")));
+			break;
+		case TSQL_FORXML_PATH: /* FOR XML PATH */
+			tsql_row_to_xml_path(state, record, element_name, binary_base64);
+			break;
+		case TSQL_FORXML_EXPLICIT:
+			/*
+			 * TODO: EXPLICIT mode is quite different from the other mode and is
+			 * not supported yet.
+			 */
+			ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("EXPLICIT mode is not supported")));
+			break;
+		default:
+			/* Invalid mode, should not happen, report internal error */
+			ereport(ERROR,
+						(errcode(ERRCODE_INTERNAL_ERROR),
+							errmsg("invalid FOR XML mode")));
+	}
+	
+	PG_RETURN_TEXT_P(cstring_to_text_with_len(state->data, state->len));
+}
+
+/*
+ * Map an SQL row to an XML element in RAW mode.
+ */
+static void
+tsql_row_to_xml_raw(StringInfo state, Datum record, const char* element_name, bool binary_base64)
+{
+	HeapTupleHeader td;
+	Oid				tupType;
+	int32			tupTypmod;
+	TupleDesc		tupdesc;
+	HeapTupleData 	tmptup;
+	HeapTuple		tuple;
+
+	td = DatumGetHeapTupleHeader(record);
+
+	/* Extract rowtype info and find a tupdesc */
+	tupType = HeapTupleHeaderGetTypeId(td);
+	tupTypmod = HeapTupleHeaderGetTypMod(td);
+	tupdesc = lookup_rowtype_tupdesc(tupType, tupTypmod);
+
+	/* Build a temporary HeapTuple control structure */
+	tmptup.t_len = HeapTupleHeaderGetDatumLength(td);
+	tmptup.t_data = td;
+	tuple = &tmptup;
+
+	/* each tuple is its own tag in raw mode */
+	appendStringInfo(state, "<%s", element_name);
+
+	/* process the tuple into attributes */
+	for (int i = 0; i < tupdesc->natts; i++)
+	{
+		char	*colname;
+		Datum	colval;
+		bool	isnull;
+		Oid 	datatype_oid;
+		Oid 	nspoid;
+		Oid 	tsql_datatype_oid;
+		char	*typename;
+		Form_pg_attribute att = TupleDescAttr(tupdesc, i);
+
+		if (att->attisdropped)
+			continue;
+
+		colname = map_sql_identifier_to_xml_name(NameStr(att->attname), true, false);
+		colval = heap_getattr(tuple, i + 1, tupdesc, &isnull);
+		datatype_oid = att->atttypid;
+
+		/* 
+		 * Below is a workaround for is_tsql_x_datatype() which does not work as expected.
+		 * We compare the datatype oid of the columns with the tsql_datatype_oid and
+		 * then specially handle some TSQL-specific datatypes.
+		 */
+		typename = SPI_gettype(tupdesc, i+1);
+		nspoid = get_namespace_oid("sys", true);
+		Assert(nspoid != InvalidOid);
+
+		tsql_datatype_oid = GetSysCacheOid2(TYPENAMENSP, Anum_pg_type_oid, CStringGetDatum(typename), ObjectIdGetDatum(nspoid));
+	
+		/*
+		 * tsql_datatype_oid can be different from datatype_oid when there are datatypes in different namespaces
+		 * but with the same name. Examples: bigint, int, etc.
+		 */
+		if (tsql_datatype_oid == datatype_oid)
```

Comment:
What about `bit` and `money` types? Do they behave differently than `for json`?

Reply:
Yeah, bit and money have a different behavior compared with `for json` - they are not coerced to boolean/numeric types and are left as-is with `for xml`